ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/RelOptHiveTable.java

```diff
@@ -333,12 +360,20 @@ public boolean isKey(ImmutableBitSet columns) {
           parentTableQualifiedName.add(parentTableName);
           qualifiedName = parentTableName;
         }
+        if (!tablesCache.containsKey(qualifiedName)) {
+          // Table doesn't exist in the cache, so we don't need to track
+          // these referential constraints. But we do need to keep track
+          // of the table in case the tableCache gets populated later, though
+          // in theory, this should never happen based on how this is called.
```

Comment:
Trying to expect the unexpected is very hard to do :)
Can't we roll some rules/conventions/etc and/or throw exception if that happens instead of trying to cover a case we don't even know will happen?
It could probably make things more straight as well

Reply:
Here's the reason why I did it this way:

RelOptHiveTable takes in a hash map containing the table name with the table object.  In my perfect world of coding, this hash map would be immutable.  It makes coding much easier when the objects are static at initialization time.

This is the proper way to code it, in my opinion.  That would make this test unnecessary.

However, this hash map can't be immutable because it is being populated while each RelOptHiveTable is being constructed.  

Once this code is rewritten, we can then remove this extra code.  But I think defensive coding here is best.  By coding it this way, the RelOptHiveTable will stay internally consistent no matter how it is called, while that table hash map stays mutable.

Reply:
I was thinking about this a little more.  I'll have to review the code to see if it's possible, but I think I can at least put in one check (that isn't foolproof) that willl prevent a user calling for an incomplete list of referential constraints.  I can do this by ensuring that a second call retrieving them doesn't have any missing.  

Of course, if the list is incomplete and a second call isn't made, then this won't be caught.

But I think will add this shortly.

Reply:
Ok, in my most recent push, I created a hash map wrapper.  This wrapper allows the caller to mark the fact that all tables have been parsed.  

The call for getting the referential constraints is now only allowed when the table map is marked as parsed so we can be assured that the list of tables used in the query is complete when fetching the referential constraints.  I think I did this with minimal intrusion on the original code, though some of that still does need rewriting.

Thanks for the review comments and making me think about this a bit more!